src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -31,24 +31,19 @@ public SolutionAssetProvider(HostWorkspaceServices services)
             _services = services;
         }
 
-        public async ValueTask GetAssetsAsync(PipeWriter pipeWriter, Checksum solutionChecksum, Checksum[] checksums, CancellationToken cancellationToken)
+        public async ValueTask<bool> TryGetAssetsAsync(PipeWriter pipeWriter, Checksum solutionChecksum, Checksum[] checksums, CancellationToken cancellationToken)
         {
             var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
             var serializer = _services.GetRequiredService<ISerializerService>();
-            var scope = assetStorage.GetScope(solutionChecksum);
 
-            SolutionAsset? singleAsset = null;
-            IReadOnlyDictionary<Checksum, SolutionAsset>? assetMap = null;
-
-            if (checksums.Length == 1)
-            {
-                singleAsset = await scope.GetAssetAsync(checksums[0], cancellationToken).ConfigureAwait(false);
-                singleAsset ??= SolutionAsset.Null;
-            }
-            else
-            {
-                assetMap = await scope.GetAssetsAsync(checksums, cancellationToken).ConfigureAwait(false);
-            }
+            // Attempt to get the scope associated with this checksum.  Note: the scope may have been released if all
+            // active outgoing operations from the inproc client were cancelled.  Even in that case it is still possible
```

Comment:
> Note: the scope may have been released if all active outgoing operations from the inproc client were cancelled

The cancellation ordering does not allow this to happen (strict requirement that this does not occur).

Reply:
Specifically, this chain of events occurs:

1. Feature one creates the async lazy, then awaits it.
2. Feature two comes in, and sees that async lazy, and awaits it.
3. The async lazy now spins up a Task to do the work, passing it a fresh CT.
4. The CTs for features one and two are cancelled.  Internally inside the AsyncLazy this goes through teh following codepaths:

First note that when both features called GetValueAsync(ct) on the async lazy, this happened:

https://github.com/dotnet/roslyn/blob/main/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy%601.cs#L351

```c#
 request.RegisterForCancellation(OnAsynchronousRequestCancelled, cancellationToken);
//
            public void RegisterForCancellation(Action<object?> callback, CancellationToken cancellationToken)
            {
                _cancellationToken = cancellationToken;
                _cancellationTokenRegistration = cancellationToken.Register(callback, this);
            }
```

So the CT passed into GetValueAsync is registered to call `OnAsynchronousRequestCancelled` once it is canceled.  Now let's look at what OnAsynchronousRequestCancelled does:

```c#
        private void OnAsynchronousRequestCancelled(object? state)
        {
            var request = (Request)state!;
            CancellationTokenSource? cancellationTokenSource = null;

            // ...

            request.Cancel();
```

So it cancels the requst unilaterally.  Let's look at that:

https://github.com/dotnet/roslyn/blob/9dc52f1a63a050f6221de460e00a2def2abca564/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy%601.cs#L587

```c#
            public void Cancel()
                => this.TrySetCanceled(_cancellationToken);
```

This explicitly and immediately sets the TCS here to the canceled state.  The TCS whose task is exactly what we return here as the result of GetValueAsync

https://github.com/dotnet/roslyn/blob/main/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy%601.cs#L358

```c#
            return request.Task;
```

So the caller of GetValueAsync will cancel out, but nothing at all has ensured that the async callback code that the async lazy was computing has actually completed at all (or can do anything to ensure it stops).